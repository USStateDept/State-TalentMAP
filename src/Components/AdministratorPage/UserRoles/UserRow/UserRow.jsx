import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { userHasPermissions } from 'utilities';
import CheckBox from 'Components/CheckBox';
import { get } from 'lodash';
import { EMPTY_FUNCTION } from 'Constants/PropTypes';
import { modifyPermission } from 'actions/userRoles';
import Skeleton from 'react-loading-skeleton';

class UserRow extends Component {
  checkPermission(permission) {
    const isEmpty = !this.props.permissionGroups.length;
    if (isEmpty) return false;

    const userPerms = this.props.permissionGroups.map(permObj => permObj.name);

    return userHasPermissions(permission, userPerms || []);
  }

  updatePermission(addRole, groupID) {
    this.props.modifyPermission(addRole, this.props.userID, groupID);
  }

  render() {
    const {
      username, name, userID, delegateRoles, isLoading,
    } = this.props;

    const tdArray = [];
    Object.keys(this.props.delegateRoles).forEach(role => (
      tdArray.push(
        <td key={get(delegateRoles, `${role}.group_name`)} className="delegateRoleCell">
          <CheckBox
            label={`Toggle ${get(delegateRoles, `${role}.group_name`)} permission.`}
            id={`${userID}-${get(delegateRoles, `${role}.group_name`)}`}
            value={this.checkPermission([get(delegateRoles, `${role}.group_name`)])}
            onCheckBoxClick={e => this.updatePermission(e, get(delegateRoles, `${role}.group_id`))}
            labelSrOnly
          />
        </td>,
      )
    ));

    return (
      <>
        {
          isLoading ?
            (
              <tr>
                <td><Skeleton /></td>
                <td><Skeleton /></td>
                <td><Skeleton /></td>
                <td><Skeleton /></td>
              </tr>
            ) :
            (
              <tr>
                <td>{username}</td>
                <td>{name}</td>
                {tdArray}
              </tr>
            )
        }
      </>
    );
  }
}

UserRow.propTypes = {
  userID: PropTypes.number.isRequired,
  username: PropTypes.string,
  name: PropTypes.string,
  permissionGroups: PropTypes.arrayOf(PropTypes.shape({})),
  delegateRoles: PropTypes.shape({}),
  modifyPermission: PropTypes.func,
  isLoading: PropTypes.bool,
};

UserRow.defaultProps = {
  username: '',
  name: '',
  permissionGroups: [],
  delegateRoles: {},
  modifyPermission: EMPTY_FUNCTION,
  onClick: EMPTY_FUNCTION,
  isLoading: false,
};

const mapStateToProps = state => ({
  modifyPermissionIsLoading: state.modifyPermissionIsLoading,
  modifyPermissionHasErrored: state.modifyPermissionHasErrored,
});

export const mapDispatchToProps = dispatch => ({
  modifyPermission: (addRole, groupID, userID) => dispatch(modifyPermission(addRole,
    groupID, userID)),
});

export default connect(mapStateToProps, mapDispatchToProps)(UserRow);
