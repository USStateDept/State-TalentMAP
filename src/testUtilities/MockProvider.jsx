import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { FlagsProvider } from 'flag';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import getFlags from '../flags';

const flags = () => getFlags();
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

export const MockTestProvider = ({ flagProps, providerProps, memoryProps, children }) => (
  <FlagsProvider flags={flags()} {...flagProps}>
    <Provider store={mockStore({})} {...providerProps}>
      <MemoryRouter {...memoryProps}>
        {children}
      </MemoryRouter>
    </Provider>
  </FlagsProvider>
);

MockTestProvider.propTypes = {
  flagProps: PropTypes.shape({}),
  providerProps: PropTypes.shape({}),
  memoryProps: PropTypes.shape({}),
  children: PropTypes.node.isRequired,
};

MockTestProvider.defaultProps = {
  flagProps: {},
  providerProps: {},
  memoryProps: {},
};

export default MockTestProvider;
