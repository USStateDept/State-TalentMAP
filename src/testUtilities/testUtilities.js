import sinon from 'sinon';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockAdapter from 'axios-mock-adapter';
import api from '../api';

// Tests all functions within a mapDispatchToProps function.
// We take two arguments, one is the entire mapDispatchToProps function, which
// can be exported from the Container/Component. The second is a config
// object with any parameters we want to pass to the function with the same name.
// For example, to pass the argument "test" to the getData() function, and "test"
// and {1} to submitData() function, we could create an object such as
// { getData: ["test"], submitData: ["test", 1] }
// Then we spy dispatch and call each property of the mapDispatchToProps
// object. Calling each property should trigger the call count on the spy,
// which we test. We reset call count between tests since we're using the
// calledOnce method.
export function testDispatchFunctions(mapDispatchToProps, config = {}) {
  it('should call all mapDispatchToProps actions', () => {
    // create spy
    const dispatchSpy = sinon.spy();
    // object with all functions
    const spiedProps = mapDispatchToProps(dispatchSpy);
    // iterate on each key
    Object.keys(spiedProps).forEach((prop) => {
      // set the prop to test
      const propToTest = spiedProps[prop];
      // set any params passed in the config object
      let params = [];
      // if there any arguments for that param, set them
      if (config[prop]) { params = config[prop]; }
      // call the prop
      propToTest(...params);
      // ensure it was called once
      sinon.assert.calledOnce(dispatchSpy);
      // reset so that called count is 0
      dispatchSpy.reset();
    });
  });
}

export function setupAsyncMocks() {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const mockAdapter = new MockAdapter(api());

  return { mockStore, mockAdapter };
}

// Form a mock adapter using a url, response, type, and callbackType.
// Spy on the mock and return the mock and the spy.
export const spyMockAdapter = ({ url = '', response = [], type = 'onGet', callbackType = 'reply' }) => {
  const mockAdapter = new MockAdapter(api());
  const spy = sinon.spy(() => response);
  const mock = () => mockAdapter[type](url)[callbackType](spy);
  return { mock, spy };
};

// Expect that some assertion was performed on a provided spy.
// Optional callback once complete.
export const expectMockWasCalled = ({ spy = () => {}, cb = () => {}, timeout = 0, assertion = 'calledOnce' }) => {
  setTimeout(() => {
    sinon.assert[assertion](spy);
    cb();
  }, timeout);
};
