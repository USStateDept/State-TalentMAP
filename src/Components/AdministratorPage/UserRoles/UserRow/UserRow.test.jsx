import { shallow } from 'enzyme';
import sinon from 'sinon';
import toJSON from 'enzyme-to-json';
import UserRow, { mapDispatchToProps } from './UserRow';
import { testDispatchFunctions } from '../../../../testUtilities/testUtilities';

describe('UserRow', () => {
  const props = {
    userID: 0,
    username: 'myTestUsername',
    name: 'my test',
    permissionGroups: [],
    delegateRoles: {
      superuser: { title: 'Super User', group_id: 1, group_name: 'superuser' },
      glossary_editors: { title: 'Glossary Editor', group_id: 2, group_name: 'glossary_editors' },
      aboutpage_editor: { title: 'About Page Editor', group_id: 3, group_name: 'aboutpage_editor' },
    },
    modifyPermission: () => [],
    updatePermission: () => [],
    checkPermission: () => [],
    onClick: () => [],
  };
  it('is defined', () => {
    const wrapper = shallow(<UserRow.WrappedComponent {...props} />);
    expect(wrapper).toBeDefined();
  });
  // TODO - revisit these tests per https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/814
  xit('calls checkPermission()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<UserRow.WrappedComponent {...props} checkPermission={spy} />);
    wrapper.instance().checkPermission();
    sinon.assert.calledOnce(spy);
  });
  // TODO - revisit these tests per https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/814
  xit('calls updatePermission() on checkbox click', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<UserRow.WrappedComponent {...props} updatePermission={spy} />);
    // wrapper.find('CheckBox').at(0).instance().onCheckBoxClick();
    wrapper.find('CheckBox').at(0).props().onCheckBoxClick();

    sinon.assert.calledOnce(spy);
  });
  // TODO - revisit these tests per https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/814
  xit('calls updatePermission()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<UserRow.WrappedComponent {...props} updatePermission={spy} />);
    wrapper.instance().updatePermission();
    sinon.assert.calledOnce(spy);
  });

  it('calls modifyPermission() from updatePermission()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<UserRow.WrappedComponent {...props} modifyPermission={spy} />);
    wrapper.instance().updatePermission();
    sinon.assert.calledOnce(spy);
  });
  // TODO - revisit these tests per https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/814
  xit('calls userHasPermissions() from checkPermission()', () => {
    const spy = sinon.spy();
    const wrapper = shallow(<UserRow.WrappedComponent {...props} userHasPermissions={spy} />);
    wrapper.instance().checkPermission();
    sinon.assert.calledOnce(spy);
  });

  it('checks the table content', () => {
    const wrapper = shallow(<UserRow.WrappedComponent {...props} />);
    expect(wrapper.find('td').at(0).props().children).toBe('myTestUsername');
    expect(wrapper.find('td').at(1).props().children).toBe('my test');
  });
  // TODO - revisit these tests per https://github.com/MetaPhase-Consulting/State-TalentMAP/pull/814
  xit('checks the td checkbox rendering', () => {
    const wrapper = shallow(<UserRow.WrappedComponent {...props} />);
    // expect(wrapper.find('td').at(2).props().children.id).toBe('0-superuser');
    expect(wrapper.find('CheckBox').at(0)).toBe('0-superuser');
    // expect(wrapper.find('CheckBox').at(1).props().children.id).toBe('0-glossary_editors');
    // expect(wrapper.find('CheckBox').at(2).props().children.id).toBe('0-aboutpage_editor');
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <UserRow.WrappedComponent {...props} />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});

describe('mapDispatchToProps', () => {
  const config = {
    modifyPermission: [1],
  };
  testDispatchFunctions(mapDispatchToProps, config);
});
