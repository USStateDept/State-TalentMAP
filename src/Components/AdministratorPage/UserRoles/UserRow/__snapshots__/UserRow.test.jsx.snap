// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRow matches snapshot 1`] = `
<Fragment>
  <tr>
    <td>
      myTestUsername
    </td>
    <td>
      my test
    </td>
    <td
      className="delegateRoleCell"
      key="superuser"
    >
      <CheckBox
        checkboxProps={Object {}}
        className=""
        disabled={false}
        id="0-superuser"
        label="Toggle superuser permission."
        labelSrOnly={true}
        name=""
        onChange={[Function]}
        onCheckBoxClick={[Function]}
        overrideLifecycle={false}
        small={false}
        title=""
        value={false}
      />
    </td>
    <td
      className="delegateRoleCell"
      key="glossary_editors"
    >
      <CheckBox
        checkboxProps={Object {}}
        className=""
        disabled={false}
        id="0-glossary_editors"
        label="Toggle glossary_editors permission."
        labelSrOnly={true}
        name=""
        onChange={[Function]}
        onCheckBoxClick={[Function]}
        overrideLifecycle={false}
        small={false}
        title=""
        value={false}
      />
    </td>
    <td
      className="delegateRoleCell"
      key="aboutpage_editor"
    >
      <CheckBox
        checkboxProps={Object {}}
        className=""
        disabled={false}
        id="0-aboutpage_editor"
        label="Toggle aboutpage_editor permission."
        labelSrOnly={true}
        name=""
        onChange={[Function]}
        onCheckBoxClick={[Function]}
        overrideLifecycle={false}
        small={false}
        title=""
        value={false}
      />
    </td>
  </tr>
</Fragment>
`;
